Implementation of vectors

guangyuz opened the pull request:
* new added bytecode: 
   - VecPack, VecUnpack, VecMut(Imm)Borrow, VecLen, VecPushBack, VecPopBack, VecSwap

* keep value type unchanged
   - haven't distinguish between different vector types, they share the same container type
   - type check is skipped at the moment (rely on the bytecode verification)

* adjust the rw_op addressing scheme:
   - extend the rw_op and add a new field 'value_ex', can write two simple value in one rw_op
   - add a header for the flattened container value to save the container length and flattened length
   - distinguish between a ref to a container and a ref to the container's field 0

* todo: 
   - the nested vectors doesn't well tested (to support the adaptive word length first)
   - read flattened length from container header instead of passing through auxiliary cell

guangyuz committed:
Support vectors in the interpreter

guangyuz committed:
fmt and bugfix

nanne007 commented:
So, you treat vector as container, and vector_ref can only ref to locals or an indexed value?

luxebeng committed:
solve conflict

guangyuz committed:
add a header element for the flattened container value

- to save the length of the flattened container elements
- to distinguish between a ref to a struct and a ref to the struct's field 0

guangyuz commented:
> So, you treat vector as container, and vector_ref can only ref to locals or an indexed value?

yes, LocalLocation and IndexedLocation are the only legal location where a vector can live persistently.

guangyuz committed:
update rust to 1.64.0 and bugfix

guangyuz committed:
add constraints for vec_pack, vec_unpack

guangyuz committed:
record vector length in the container header

guangyuz committed:
add constraints for vec_borrow

guangyuz committed:
add constraints for vec_push_back

guangyuz committed:
add constraints for vec_pop_back

guangyuz committed:
constrains vec_swap

guangyuz committed:
refactor and fix clippy

guangyuz committed:
support vector in globals

luxebeng committed:
support value with None

guangyuz committed:
optimize vec_swap constraints
